File: ports/dotnet/src/RussianDocs.DocumentProcessing/ViewModel/Payload.cs
using System.Text.Json.Serialization;

namespace RussianDocs.DocumentProcessing.ViewModel;

/// <summary>
/// The view model — the detail response the SPA reads.
///
/// <para>
/// **Every JSON name is written by hand and nothing is omitted.** No naming policy, and
/// <c>JsonIgnoreCondition.Never</c> on the serialiser: the SPA reads about sixty named fields, and a
/// missing key is a real defect rather than a tolerance question — it makes a page render blank. A
/// policy that gets fifty-nine names right is worse than none, because the one it misses looks like a
/// typo somewhere else entirely.
/// </para>
///
/// <para>
/// Nullable properties are nullable ON PURPOSE. "Absent" must serialise as <c>null</c>, not as 0 or
/// "" — the SPA distinguishes a field that was not read from one that read as empty.
/// </para>
///
/// <para>
/// Exactly FOURTEEN top-level keys when debug is off. The count is asserted by the harness, because
/// an extra key is as much a contract break as a missing one.
/// </para>
/// </summary>
public sealed class Payload
{
    [JsonPropertyName("doc_type")] public string? DocType { get; set; }
    [JsonPropertyName("doc_type_base")] public string? DocTypeBase { get; set; }
    [JsonPropertyName("doc_type_era")] public string? DocTypeEra { get; set; }
    [JsonPropertyName("recognised")] public bool Recognised { get; set; }
    [JsonPropertyName("device")] public string? Device { get; set; }
    [JsonPropertyName("canvas")] public Canvas Canvas { get; set; } = new();
    [JsonPropertyName("coord_space")] public string CoordSpace { get; set; } = "canvas";
    [JsonPropertyName("coord_space_note")] public string CoordSpaceNote { get; set; } = "";
    [JsonPropertyName("boxes")] public List<Box> Boxes { get; set; } = [];
    [JsonPropertyName("fields")] public List<Field> Fields { get; set; } = [];
    [JsonPropertyName("ocr")] public Dictionary<string, string> Ocr { get; set; } = [];
    [JsonPropertyName("quality")] public Dictionary<string, object> Quality { get; set; } = [];
    [JsonPropertyName("timings")] public Dictionary<string, double> Timings { get; set; } = [];

    /// <summary>Null for every type except INTPASSPORTADDR, which this port does not implement.</summary>
    [JsonPropertyName("address")] public Address? Address { get; set; }

    /// <summary>
    /// Only present under <c>?include=debug</c> — the one key that IS omitted when absent, because
    /// the reference omits it rather than sending null.
    /// </summary>
    [JsonPropertyName("debug")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public Debug? Debug { get; set; }
}

public sealed class Canvas
{
    [JsonPropertyName("width")] public int? Width { get; set; }
    [JsonPropertyName("height")] public int? Height { get; set; }

    /// <summary>
    /// True when the pipeline short-circuited and there is no corrected canvas.
    ///
    /// <para>
    /// A separate flag rather than inferring it from null dimensions, because the UI must be able to
    /// say "this is the original upload" rather than silently drawing boxes on the wrong image.
    /// </para>
    /// </summary>
    [JsonPropertyName("is_fallback")] public bool IsFallback { get; set; }
}

public sealed class Box
{
    [JsonPropertyName("id")] public string Id { get; set; } = "";
    [JsonPropertyName("label")] public string Label { get; set; } = "";
    [JsonPropertyName("display")] public string Display { get; set; } = "";

    /// <summary><c>"text"</c> or <c>"visual"</c> — Face and Signature are the visual ones.</summary>
    [JsonPropertyName("kind")] public string Kind { get; set; } = "text";

    [JsonPropertyName("x1")] public int? X1 { get; set; }
    [JsonPropertyName("y1")] public int? Y1 { get; set; }
    [JsonPropertyName("x2")] public int? X2 { get; set; }
    [JsonPropertyName("y2")] public int? Y2 { get; set; }
    [JsonPropertyName("conf")] public double? Conf { get; set; }
    [JsonPropertyName("cls")] public int? Cls { get; set; }

    /// <summary>
    /// The field's text, on the box that OWNS it.
    ///
    /// <para>
    /// A field can be detected more than once — the internal passport prints its number twice — and
    /// the library discards the per-detection text, so only the highest-confidence box carries the
    /// value. The others get <c>ambiguous: true</c> instead.
    /// </para>
    /// </summary>
    [JsonPropertyName("text")] public string? Text { get; set; }

    /// <summary>True on a duplicate detection of an OCR'd field that does NOT own the text.</summary>
    [JsonPropertyName("ambiguous")] public bool Ambiguous { get; set; }
}

public sealed class Field
{
    [JsonPropertyName("name")] public string Name { get; set; } = "";
    [JsonPropertyName("display")] public string Display { get; set; } = "";
    [JsonPropertyName("value")] public string? Value { get; set; }
    [JsonPropertyName("script")] public string Script { get; set; } = "ru";
    [JsonPropertyName("conf")] public double? Conf { get; set; }

    /// <summary>
    /// Every box belonging to this field.
    ///
    /// <para>
    /// An ARRAY, not one id: split fields produce several boxes and duplicated fields produce more.
    /// This is what lets the UI highlight all of a field's boxes at once — linking by label string
    /// would be ambiguous exactly where it matters.
    /// </para>
    /// </summary>
    [JsonPropertyName("box_ids")] public List<string> BoxIds { get; set; } = [];
}

public sealed class Debug
{
    [JsonPropertyName("doc_outline")] public DocOutline DocOutline { get; set; } = new();
}

public sealed class DocOutline
{
    /// <summary>
    /// Always <c>"prewarp"</c>.
    ///
    /// <para>
    /// The document contour is in the space BEFORE the perspective correction, so it must not be
    /// drawn on the canvas. Tagging it explicitly is the only thing stopping a UI from doing so.
    /// </para>
    /// </summary>
    [JsonPropertyName("coord_space")] public string CoordSpace { get; set; } = "prewarp";

    [JsonPropertyName("polygon")] public List<List<int[]>>? Polygon { get; set; }
}

public sealed class Address
{
    [JsonPropertyName("aligned")] public bool Aligned { get; set; }
    [JsonPropertyName("lines")] public List<AddressLine> Lines { get; set; } = [];
}

/// <summary>
/// One address line. Declared but never produced by this port.
///
/// <para>
/// The type exists because the contract includes it and an integrator needs the shape; INTPASSPORTADDR
/// is blocked on an anonymised sample, not on code. An omitted type reads as an oversight that the
/// next port would invent differently.
/// </para>
/// </summary>
public sealed class AddressLine
{
    [JsonPropertyName("id")] public string Id { get; set; } = "";
    [JsonPropertyName("kind")] public string? Kind { get; set; }
    [JsonPropertyName("text")] public string? Text { get; set; }
    [JsonPropertyName("p_handwritten")] public double? PHandwritten { get; set; }
    [JsonPropertyName("obbox")] public Obbox? Obbox { get; set; }
}

public sealed class Obbox
{
    [JsonPropertyName("cx")] public double? Cx { get; set; }
    [JsonPropertyName("cy")] public double? Cy { get; set; }
    [JsonPropertyName("w")] public double? W { get; set; }
    [JsonPropertyName("h")] public double? H { get; set; }
    [JsonPropertyName("angle_rad")] public double? AngleRad { get; set; }
    [JsonPropertyName("conf")] public double? Conf { get; set; }
    [JsonPropertyName("label")] public string? Label { get; set; }
}

File: ports/java/docproc/src/main/kotlin/net/russiandocs/docproc/viewmodel/Payload.kt
package net.russiandocs.docproc.viewmodel

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement

/**
 * The view model — the detail response the SPA reads.
 *
 * **Every JSON name is written by hand and nothing is omitted.** No naming policy, and `explicitNulls = true`
 * on the serialiser: the SPA reads about sixty named fields, and a missing key is a real defect rather than a
 * tolerance question — it makes a page render blank. A policy that gets fifty-nine names right is worse than
 * none, because the one it misses looks like a typo somewhere else entirely.
 *
 * **kotlinx.serialization OMITS nulls by default**, which is the exact opposite of what this contract needs
 * (J-04). Nullable properties are nullable ON PURPOSE: "absent" must serialise as `null`, not vanish and not
 * become 0 or "" — the SPA distinguishes a field that was not read from one that read as empty.
 *
 * Exactly FOURTEEN top-level keys when debug is off. The count is asserted by the harness, because an extra
 * key is as much a contract break as a missing one.
 */
@Serializable
public data class Payload(
    @SerialName("doc_type") val docType: String? = null,
    @SerialName("doc_type_base") val docTypeBase: String? = null,
    @SerialName("doc_type_era") val docTypeEra: String? = null,
    @SerialName("recognised") val recognised: Boolean = false,
    @SerialName("device") val device: String? = null,
    @SerialName("canvas") val canvas: Canvas = Canvas(),
    @SerialName("coord_space") val coordSpace: String = "canvas",
    @SerialName("coord_space_note") val coordSpaceNote: String = "",
    @SerialName("boxes") val boxes: List<Box> = emptyList(),
    @SerialName("fields") val fields: List<Field> = emptyList(),
    @SerialName("ocr") val ocr: Map<String, String> = emptyMap(),
    /**
     * Heterogeneous by contract — `good`/`bad` for glare and blur, `REAL`/`FAKE` for spoofing, and a NUMBER
     * for DocConf. A [JsonElement] rather than a typed map, because no single Kotlin type expresses that and
     * normalising it would change the wire format.
     */
    @SerialName("quality") val quality: Map<String, JsonElement> = emptyMap(),
    @SerialName("timings") val timings: Map<String, Double> = emptyMap(),
    /** Null for every type except INTPASSPORTADDR, which this port does not implement. */
    @SerialName("address") val address: Address? = null,
    /**
     * Only present under `?include=debug` — the one key that IS omitted when absent, because the reference
     * omits it rather than sending null. The builder therefore emits two different payload shapes rather than
     * relying on a serialiser setting that would apply to every field.
     */
    @SerialName("debug") val debug: Debug? = null,
)

@Serializable
public data class Canvas(
    @SerialName("width") val width: Int? = null,
    @SerialName("height") val height: Int? = null,
    /**
     * True when recognition short-circuited and there is no corrected canvas.
     *
     * The SPA then shows the original upload and hides the overlay, because drawing canvas-space boxes on the
     * original would be wrong by the whole perspective transform.
     */
    @SerialName("is_fallback") val isFallback: Boolean = false,
)

@Serializable
public data class Box(
    @SerialName("id") val id: String = "",
    @SerialName("label") val label: String = "",
    @SerialName("display") val display: String = "",
    /** `"text"` or `"visual"` — Face and Signature are the visual ones. */
    @SerialName("kind") val kind: String = "text",
    @SerialName("x1") val x1: Int? = null,
    @SerialName("y1") val y1: Int? = null,
    @SerialName("x2") val x2: Int? = null,
    @SerialName("y2") val y2: Int? = null,
    @SerialName("conf") val conf: Double? = null,
    @SerialName("cls") val cls: Int? = null,
    @SerialName("text") val text: String? = null,
    /** True on a duplicate detection of an OCR'd field that does NOT own the text. */
    @SerialName("ambiguous") val ambiguous: Boolean = false,
)

@Serializable
public data class Field(
    @SerialName("name") val name: String = "",
    @SerialName("display") val display: String = "",
    @SerialName("value") val value: String? = null,
    @SerialName("script") val script: String = "ru",
    @SerialName("conf") val conf: Double? = null,
    /**
     * Every box carrying this label, in detection order.
     *
     * A list rather than one id: a field can be detected twice — the internal passport prints its series in
     * two places — and split fields like `Birth_place_ru` legitimately span several boxes.
     */
    @SerialName("box_ids") val boxIds: List<String> = emptyList(),
)

@Serializable
public data class Debug(
    @SerialName("doc_outline") val docOutline: DocOutline = DocOutline(),
)

@Serializable
public data class DocOutline(
    /**
     * `prewarp`, and the tag is load-bearing: these coordinates are in the space BEFORE the perspective
     * correction, so drawing them on the canvas would be wrong by the whole transform.
     */
    @SerialName("coord_space") val coordSpace: String = "prewarp",
    @SerialName("polygon") val polygon: List<List<List<Int>>>? = null,
)

@Serializable
public data class Address(
    @SerialName("aligned") val aligned: Boolean = false,
    @SerialName("lines") val lines: List<AddressLine> = emptyList(),
)

@Serializable
public data class AddressLine(
    @SerialName("id") val id: String = "",
    @SerialName("kind") val kind: String? = null,
    @SerialName("text") val text: String? = null,
    @SerialName("p_handwritten") val pHandwritten: Double? = null,
    @SerialName("obbox") val obbox: Obbox? = null,
)

@Serializable
public data class Obbox(
    @SerialName("cx") val cx: Double? = null,
    @SerialName("cy") val cy: Double? = null,
    @SerialName("w") val w: Double? = null,
    @SerialName("h") val h: Double? = null,
    @SerialName("angle_rad") val angleRad: Double? = null,
    @SerialName("conf") val conf: Double? = null,
    @SerialName("label") val label: String? = null,
)
